tools/class-redirector/class-redirector/Rewriter.cs

```diff
@@ -0,0 +1,258 @@
+			PatchMethods (cl, class_ptr, classPtrField);
+		}
+
+		void PatchMethods (TypeDefinition cl, FieldDefinition classPtr, FieldDefinition classPtrField)
+		{
+			foreach (var method in cl.Methods) {
+				PatchMethod (method, classPtr, classPtrField);
+			}
+		}
+
+		void PatchMethod (MethodDefinition method, FieldDefinition classPtr, FieldDefinition classPtrField)
+		{
+			foreach (var instr in method.Body.Instructions) {
+				if (instr.OpCode == OpCodes.Ldsfld && instr.Operand == classPtr) {
+					instr.Operand = classPtrField;
+				}
+			}
+		}
+
+		void RemoveCCtorInit (TypeDefinition cl, FieldDefinition class_ptr)
+		{
+			var cctor = cl.Methods.FirstOrDefault (m => m.Name == ".cctor");
+			if (cctor is null)
+				return; // no static init - should never happen, but we can deal.
+
+			var il = cctor.Body.GetILProcessor ();
+			Instruction? stsfld = null;
+			int i = 0;
+			for (; i < il.Body.Instructions.Count; i++) {
+				var instr = il.Body.Instructions [i];
+				// look for
+				// stsfld class_ptr
+				if (instr.OpCode == OpCodes.Stsfld && instr.Operand == class_ptr) {
+					stsfld = instr;
+					break;
+				}
+			}
+			if (stsfld is null)
+				return;
+
+			// if we see:
+			// ldstr "any"
+			// call ObjCRuntime.GetClassHandle
+			// stsfld class_ptr
+			// Then we can remove all of those instructions
+
+			var isGetClassHandle = IsGetClassHandle (il, i - 1);
+			var isLdStr = IsLdStr (il, i - 2);
+
+			if (isGetClassHandle && isLdStr) {
+				il.RemoveAt (i);
+				il.RemoveAt (i - 1);
+				il.RemoveAt (i - 2);
+			} else if (isGetClassHandle) {
+				// don't know how the string got on the stack, so at least get rid of the
+				// call to GetClassHandle by nopping it out. This still leaves the string on
+				// the stack, so pop it.
+				il.Replace (il.Body.Instructions [i - 1], Instruction.Create (OpCodes.Nop));
+				// can't remove all three, so just pop the IntPtr.
+				il.Replace (il.Body.Instructions [i], Instruction.Create (OpCodes.Pop));
+			} else {
+				// can't remove all three, so just pop the IntPtr.
+				il.Replace (il.Body.Instructions [i], Instruction.Create (OpCodes.Pop));
+			}
+
+			// if we're left with exactly 1 instruction and it's a return,
+			// then we can get rid of the entire method
+			if (cctor.Body.Instructions.Count == 1) {
+				if (cctor.Body.Instructions.Last ().OpCode == OpCodes.Ret)
+					cl.Methods.Remove (cctor);
+			}
+
+			// if we're left with exactly 2 instructions and the first is a no-op
+			// and the last is a return, then we can get rid of the entire method
+			if (cctor.Body.Instructions.Count == 2) {
+				if (cctor.Body.Instructions.Last ().OpCode == OpCodes.Ret &&
+					cctor.Body.Instructions.First ().OpCode == OpCodes.Nop)
+					cl.Methods.Remove (cctor);
+			}
+		}
+
+		bool IsGetClassHandle (ILProcessor il, int index)
+		{
+			if (index < 0)
+				return false;
+			var instr = il.Body.Instructions [index]!;
+			var operand = instr.Operand?.ToString () ?? "";
```

Comment:
You're calling ToString on every instruction, which is unnecessary. You only need to call ToString on `call` instructions, which would be _much_ more performant. This would mean moving the `instr.OpCode == OpCodes.Call` check to before the ToString call.